source/agora/test/Base.d

```diff
@@ -1028,3 +1034,79 @@ public bool containSameBlocks (APIS)(APIS nodes, size_t height)
 
     return true;
 }
+
+
+/*******************************************************************************
+
+    Generate a genesis block.
+
+    For each key pair, a freeze transaction and an Enrollment
+    will be created and added to the generated Genesis block.
+    The first transaction is the `UnitTestGenesisTransaction`
+    The minimum number of enrollments is Enrollment.MinValidatorCount.
+
+    Params:
+        key_pairs = key pairs for signing enrollments with
+
+    Returns:
+        The genesis block
+
+*******************************************************************************/
+
+private immutable(Block) makeGenesisBlock (in KeyPair[] key_pairs)
+{
+    import agora.common.Serializer;
+
+    // 1 payment tx, the rest are freeze txs
+    Transaction[] txs;
+    txs ~= UnitTestGenesisTransaction.serializeFull.deserializeFull!Transaction;
+    Enrollment[] enrolls;
+
+    foreach (idx; 0 .. max(Enrollment.MinValidatorCount, key_pairs.length))
+    {
+        auto key_pair = (idx < key_pairs.length)
+                        ? key_pairs[idx]
+                        : KeyPair.random;
```

Comment:
Like I mentioned in the other PR, this code is a workaround for the tests which would break because of the new protocol change.

We could do the following:

- Implement validation and set the MinValidatorCount to 1. This means existing tests will still be correct.
- Fix the tests to make them more generic and so they always spawn at least `MinValidatorCount` Validators. That means `key_pairs` here will always be at least `>= MinValidatorCount `. We will probably rename `NetworkTopology.OneValidator` to `NetworkTopology.MinimumValidators`, so for example if we want 10 nodes, 2 would automatically become Validators and the rest FullNodes.
- Increase `MinValidatorCount` to a number that we define later. Maybe we will decide that 3 is a good number, or 4.

The work that I listed here can be done independently, one after the other.

Reply:
I saw your review on #877 and moved the code here.
I understand all your suggestions. Thank you for your good suggestion.
I will proceed with the three suggestions in order.